src/payjoin/app.rs

```diff
@@ -0,0 +1,214 @@
+            //     Some(result) => Ok(result.allowed),
+            //     None => Err(Error::Server(
+            //         anyhow!("No mempool results returned on broadcast check").into(),
+            //     )),
+            // }
+            Ok(true)
+        })?;
+        tracing::trace!("check1");
+
+        use bdk::bitcoin::{Script, ScriptBuf};
+        use std::sync::mpsc;
+        use std::sync::mpsc::Sender;
+        use std::sync::Arc;
+        use std::sync::Mutex;
+
+        // Create a channel
+        let (tx, rx) = mpsc::channel::<(ScriptBuf, Sender<Result<bool, ApplicationError>>)>();
+        let tx = Arc::new(Mutex::new(tx));
+
+        let app = self.app.clone();
+        let profile = self.profile.clone();
+        let network = network.clone();
+
+        // Receive Check 2: receiver can't sign for proposal inputs
+        tokio::spawn(async move {
+            while let Ok((input, response_tx)) = rx.recv() {
+                let result = check_script_owned(&app, &profile, network, &input).await;
+                let _ = response_tx.send(result);
+            }
+        });
+
+        async fn check_script_owned(
+            app: &App,
+            profile: &Profile,
+            network: bitcoin::Network,
+            input: &Script,
+        ) -> Result<bool, ApplicationError> {
+            let address = bitcoin::BdkAddress::from_script(&input, network)?;
+            match app
+                .find_address(&profile.clone(), address.to_string())
+                .await
+            {
+                Err(ApplicationError::AddressError(e)) => Err(ApplicationError::AddressError(e)),
+                Err(_) => Ok(false), // good address script, but not found
+                Ok(_) => Ok(true),
+            }
+        }
+
+        let proposal = proposal.check_inputs_not_owned(|input| {
+            let (response_tx, response_rx) = mpsc::channel::<Result<bool, ApplicationError>>();
+            {
+                let tx = tx.lock().expect("lock"); // TODO Handle lock error if needed
+                tx.send((input.to_owned(), response_tx))
+                    .map_err(|e| Error::Server(e.into()))?;
+            }
+            let recv_res = response_rx.recv().map_err(|e| Error::Server(e.into()))?;
+            Ok(recv_res.map_err(|e| Error::Server(e.into()))?)
+        })?;
+
+        tracing::trace!("check2");
+        // Receive Check 3: receiver can't sign for proposal inputs
+        let proposal = proposal.check_no_mixed_input_scripts()?;
+        tracing::trace!("check3");
+
+        // Receive Check 4: have we seen this input before? More of a check for non-interactive i.e. payment processor receivers.
+        let payjoin = proposal.check_no_inputs_seen_before(|input| {
+            // TODO implement input_seen_before database check
+            // Ok(!self.insert_input_seen_before(*input).map_err(|e| Error::Server(e.into()))?)
+            Ok(false)
+        })?;
+        tracing::trace!("check4");
```

Comment:
Since clients can POST payjoin requests which return proposal_psbt including bria utxos, they should be limited with this check. If an adversary were to submit many requests and never sign the proposal_psbt then they would be able to see many utxos within bria's wallets. This is mitigated with this check to only reply with bria utxos for request inputs the first time they are seen and also by broadcasting original_psbt included in the POST request after some timeout if the proposal_psbt does is not signed and broadcast.

Reply:
This will require a new repo that keeps track of seen inputs

Reply:
Anything Utxo related (selection, filter, tagging) should be put in the utxo module. There is already an entity for utxo.